components/brave_wallet/browser/bitcoin/bitcoin_rpc.cc

```diff
@@ -0,0 +1,251 @@
+
+#include "base/command_line.h"
+#include "base/ranges/algorithm.h"
+#include "base/strings/string_util.h"
+#include "brave/components/brave_wallet/browser/brave_wallet_utils.h"
+#include "brave/components/brave_wallet/common/brave_wallet.mojom-forward.h"
+#include "brave/components/brave_wallet/common/switches.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+
+namespace {
+
+net::NetworkTrafficAnnotationTag GetNetworkTrafficAnnotationTag() {
+  return net::DefineNetworkTrafficAnnotation("bitcoin_rpc", R"(
+      semantics {
+        sender: "Bitcoin RPC"
+        description:
+          "This service is used to communicate with Bitcoin nodes "
+          "on behalf of the user interacting with the native Brave wallet."
+        trigger:
+          "Triggered by uses of the native Brave wallet."
+        data:
+          "Bitcoin JSON RPC response bodies."
+        destination: WEBSITE
+      }
+      policy {
+        cookies_allowed: NO
+        setting:
+          "You can enable or disable this feature on chrome://flags."
+        policy_exception_justification:
+          "Not implemented."
+      }
+    )");
+}
+
+bool IsAsciiAlphaNumeric(const std::string& str) {
+  return base::ranges::all_of(str, &base::IsAsciiAlphaNumeric<char>);
+}
+
+const GURL MakeGetChainHeightUrl(const GURL& base_url) {
+  GURL::Replacements replacements;
+  const std::string path =
+      base_url.path() + base::JoinString({"blocks", "tip", "height"}, "/");
+  replacements.SetPathStr(path);
+
+  return base_url.ReplaceComponents(replacements);
+}
+
+const GURL MakeAddressHistoryUrl(const GURL& base_url,
+                                 const std::string& address,
+                                 const std::string& last_seen_txid) {
+  if (!IsAsciiAlphaNumeric(address) || !IsAsciiAlphaNumeric(last_seen_txid)) {
+    return GURL();
+  }
+
+  GURL::Replacements replacements;
+  std::string path =
+      base_url.path() +
+      base::JoinString({"address", address, "txs", "chain"}, "/");
+  if (!last_seen_txid.empty()) {
+    path += "/" + last_seen_txid;
+  }
+  replacements.SetPathStr(path);
+
+  return base_url.ReplaceComponents(replacements);
+}
+
+const GURL MakePostTransactionUrl(const GURL& base_url) {
+  GURL::Replacements replacements;
+  const std::string path = base_url.path() + "tx";
+  replacements.SetPathStr(path);
+
+  return base_url.ReplaceComponents(replacements);
+}
+
+std::string BitcoinRpcHost() {
+  return base::CommandLine::ForCurrentProcess()->GetSwitchValueASCII(
+      brave_wallet::switches::kBitcoinRpcHost);
+}
+
+GURL BaseRpcUrl(const std::string& network_id) {
+  CHECK(brave_wallet::IsBitcoinNetwork(network_id));
+
+  std::string host = BitcoinRpcHost();
+  std::string path = network_id == brave_wallet::mojom::kBitcoinMainnet
+                         ? "/api/"
+                         : "/testnet/api/";
+
+  return GURL("https://" + host + path);
```

Comment:
base::StrCat instead of + for all use cases in this file.

Reply:
https://github.com/brave/brave-core/pull/17978/commits/72d48e56e6778c7480f03b8d5fd63175c33eadf5